### PR TITLE
Remove ReferenceCounted and add close() to ClientTransportFactory. Fixes #927

### LINKED ISSUE #927: ReferenceCounted is useless
777e928 introduced reference counting for channel builders, but nothing ever does a retain(), making it useless.

I thought the original reason to propose having reference counting was so that transports would `retain()` their factory and then `ChannelImpl` would `release()` the factory immediately on `shutdown()`. As the transports shutdown they would each `release()` and any shared resources would naturally be released.

@nmittler, FYI

### PR DESCRIPTION
**Pull request by buchgr**
Maybe I overengineered, considering that `ClientTransportFactory` is internal. Unsure. You tell me :-).

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#Wed Mar 02 09:04:21 PST 2016
+#Wed Mar 16 22:14:48 CET 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.11-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.11-all.zip
```
Comment (ejona86) [inline gradle/wrapper/gradle-wrapper.properties]: Unrelated?

Comment (buchgr) [inline gradle/wrapper/gradle-wrapper.properties]: yep sorry. didn't notice. I think that was my intellij gradle plugin. will revert.

File: core/src/main/java/io/grpc/inprocess/InProcessChannelBuilder.java [hunk 2]
```diff
@@ -80,21 +80,30 @@ protected ClientTransportFactory buildTransportFactory() {
     return new InProcessClientTransportFactory(name);
   }
 
-  private static class InProcessClientTransportFactory extends AbstractReferenceCounted
-          implements ClientTransportFactory {
+  /**
+   * Creates InProcess transports. Exposed for internal use, as it should be private.
+   */
+  @Internal
+  protected static final class InProcessClientTransportFactory implements ClientTransportFactory {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/inprocess/InProcessChannelBuilder.java]: package-protected? Ditto for the other cases.

**Comment by ejona86** [comment]
We've generally not been bothering with checks like the close, since there is only one caller and multiple implementations; it's easier to just fix the one caller instead of adding checks to each implementation. But this can be fine.

@buchgr, this LGTM after reducing visibility of the factories to package-private.

**Comment by buchgr** [comment]
thanks Eric! cherry-picked (3c68c053f79470f63646c).
